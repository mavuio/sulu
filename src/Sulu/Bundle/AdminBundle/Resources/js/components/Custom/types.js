// @flow
import type {ElementRef} from 'react';

export type CustomProps<T: ?string | ?number> = {|
    value: ?T,
|};
