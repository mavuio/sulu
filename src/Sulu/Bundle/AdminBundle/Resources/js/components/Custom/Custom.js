// @flow
import React, {Fragment} from 'react';
import type {CustomProps} from './types';


const $script = require("scriptjs")



export default class Custom<T: ?string | ?number> extends React.PureComponent<CustomProps<T>> {


    constructor(props) {
        super(props);
        this.state = {Component: null};

        $script("/tab_bar/mavu-tabbar.js", () => {
            this.setState({
                Component: ({ data, children, ...props}) => React.createElement(
                    'mavu-tabbar',
                    {...props, data: JSON.stringify(data), ref: elem => this[name] = elem },
                    children
                )
            });
            // this[name].addEventListener('sendData', event => {
            //     const payload = event.detail
            //     payload && payload.length && sendData(payload[0])
            // });
        });
      }


    render() {
        const {
            value,
        } = this.props;

        if(this.state.Component)
        {
            return (<this.state.Component  {...this.props} />);
        } else {
            return (<div>loading...</div>)
        }
    }
}
