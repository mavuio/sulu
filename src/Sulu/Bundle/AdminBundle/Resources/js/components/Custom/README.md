The custom component can be used to get custom from the user in the same way as with the native browser custom.

```javascript
const [value, setValue] = React.useState('');

<Custom value={value} onChange={setValue} />
```

Beneath attributes known from the native custom, it provides properties to style the the custom.

```javascript
const [value, setValue] = React.useState('');

<Custom icon="fa-key" type="password" placeholder="Password" value={value} onChange={setValue} />
```

It also offers a `headline` prop, which allows to use distinguish more important fields from others.

```javascript
const [value, setValue] = React.useState('');

<Custom icon="fa-key" headline={true} value={value} onChange={setValue} />
```

When setting the `valid` prop to `false` it will mark the field as invalid. The following example shows an custom field
that needs to contain some text.

```javascript
const [value, setValue] = React.useState('');
const [valid, setValid] = React.useState(false);

const onChange = (newValue) => {
    setValue(newValue);
    setValid(!!newValue);
};

<Custom valid={valid} value={value} onChange={onChange} />
```

In addition to that the `onBlur` callback will be executed when `Custom` components loses the focus.

```javascript
const [value, setValue] = React.useState('');

<Custom value={value} onChange={setValue} onBlur={() => alert('Focus lost!')} />
```

The component also supports limiting the amount of characters typed into it.

```javascript
const [value, setValue] = React.useState('');

<Custom value={value} maxCharacters={5} onChange={setValue} />
```

It even supports limiting the amount of segments, which is super useful if a specific number of keywords should be
delimited e.g. by a comma.

```javascript
const [value, setValue] = React.useState('');

<Custom value={value} maxSegments={5} segmentDelimiter="," onChange={setValue} />
```
