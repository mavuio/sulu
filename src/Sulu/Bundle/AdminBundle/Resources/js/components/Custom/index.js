// @flow
import Custom from './Custom';
import type {CustomProps} from './types';

export default Custom;
export type {CustomProps};
