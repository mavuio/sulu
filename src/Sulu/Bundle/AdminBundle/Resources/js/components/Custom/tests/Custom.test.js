// @flow
import React from 'react';
import {render, mount, shallow} from 'enzyme';
import Custom from '../Custom';

jest.mock('../../../utils/Translator', () => ({
    translate: jest.fn((key) => key),
}));

test('Custom should render', () => {
    const onChange = jest.fn();
    expect(render(<Custom onBlur={jest.fn()} onChange={onChange} value="My value" />)).toMatchSnapshot();
});

test('Custom should render with autocomplete off', () => {
    const onChange = jest.fn();
    expect(
        render(<Custom autocomplete="off" disabled={true} onChange={onChange} value="My value" />)
    ).toMatchSnapshot();
});

test('Custom should render as headline', () => {
    expect(render(<Custom headline={true} onChange={jest.fn()} value="My value" />)).toMatchSnapshot();
});

test('Custom should render with invalid value', () => {
    const onChange = jest.fn();
    expect(render(<Custom onBlur={jest.fn()} onChange={onChange} valid={false} value="My value" />)).toMatchSnapshot();
});

test('Custom should render when disabled', () => {
    const onChange = jest.fn();
    expect(render(<Custom disabled={true} onChange={onChange} value="My value" />)).toMatchSnapshot();
});

test('Custom should render with icon', () => {
    const onChange = jest.fn();
    expect(render(<Custom icon="su-pen" onBlur={jest.fn()} onChange={onChange} value="My value" />)).toMatchSnapshot();
});

test('Custom should render with custommode', () => {
    const onChange = jest.fn();
    expect(render(
        <Custom customMode="numeric" onBlur={jest.fn()} onChange={onChange} value="My value" />
    )).toMatchSnapshot();
});

test('Custom should render with type', () => {
    const onChange = jest.fn();
    expect(render(
        <Custom onBlur={jest.fn()} onChange={onChange} type="password" value="My value" />
    )).toMatchSnapshot();
});

test('Custom should render with placeholder', () => {
    const onChange = jest.fn();
    expect(render(
        <Custom onBlur={jest.fn()} onChange={onChange} placeholder="My placeholder" value="My value" />
    )).toMatchSnapshot();
});

test('Custom should render with value', () => {
    const onChange = jest.fn();
    expect(render(<Custom onBlur={jest.fn()} onChange={onChange} value="My value" />)).toMatchSnapshot();
});

test('Custom should render undefined value as empty string', () => {
    const onChange = jest.fn();
    expect(render(<Custom onBlur={jest.fn()} onChange={onChange} value={undefined} />)).toMatchSnapshot();
});

test('Custom should render with a character counter', () => {
    expect(render(<Custom maxCharacters={2} onBlur={jest.fn()} onChange={jest.fn()} value="asdf" />)).toMatchSnapshot();
});

test('Custom should render with a segment counter', () => {
    expect(render(
        <Custom
            maxSegments={3}
            onBlur={jest.fn()}
            onChange={jest.fn()}
            segmentDelimiter=","
            value="keyword1, keyword2"
        />
    )).toMatchSnapshot();
});

test('Custom should call the callback when the custom changes', () => {
    const onChange = jest.fn();
    const custom = shallow(<Custom onBlur={jest.fn()} onChange={onChange} value="My value" />);
    const event = {currentTarget: {value: 'my-value'}};
    custom.find('custom').simulate('change', event);
    expect(onChange).toHaveBeenCalledWith('my-value', event);
});

test('Custom should call the callback with undefined if the custom value is removed', () => {
    const onChange = jest.fn();
    const custom = shallow(<Custom onBlur={jest.fn()} onChange={onChange} value="My value" />);
    const event = {currentTarget: {value: ''}};
    custom.find('custom').simulate('change', event);
    expect(onChange).toHaveBeenCalledWith(undefined, event);
});

test('Custom should call the callback when icon was clicked', () => {
    const onChange = jest.fn();
    const handleIconClick = jest.fn();
    const custom = mount(<Custom icon="su-pen" onChange={onChange} onIconClick={handleIconClick} value="My value" />);
    custom.find('Icon').simulate('click');
    expect(handleIconClick).toHaveBeenCalled();
});

test('Custom should call the given focus callback', () => {
    const onFocusSpy = jest.fn();

    const custom = mount(<Custom icon="su-pen" onChange={jest.fn()} onFocus={onFocusSpy} value="My value" />);

    expect(onFocusSpy).not.toHaveBeenCalled();
    custom.find('custom').simulate('focus');
    expect(onFocusSpy).toHaveBeenCalled();
});

test('Custom should render with a loader', () => {
    const onChange = jest.fn();
    expect(render(<Custom loading={true} onBlur={jest.fn()} onChange={onChange} value={undefined} />)).toMatchSnapshot();
});

test('Custom should render collapsed', () => {
    expect(render(<Custom collapsed={true} onChange={jest.fn()} value={undefined} />)).toMatchSnapshot();
});

test('Custom should render append container when onClearClick callback is provided', () => {
    expect(render(<Custom onChange={jest.fn()} onClearClick={jest.fn()} value={undefined} />)).toMatchSnapshot();
});

test('Custom should render append container with icon when onClearClick callback is provided and value is set', () => {
    expect(render(<Custom onChange={jest.fn()} onClearClick={jest.fn()} value="test" />)).toMatchSnapshot();
});

test('Custom should should call the callback when clear icon was clicked', () => {
    const onClearClick = jest.fn();
    const custom = mount(<Custom onChange={jest.fn()} onClearClick={onClearClick} value="My value" />);
    custom.find('Icon').simulate('click');
    expect(onClearClick).toHaveBeenCalled();
});

test('Custom should render with dark skin', () => {
    expect(
        render(<Custom icon="su-pen" onChange={jest.fn()} onClearClick={jest.fn()} skin="dark" value={undefined} />)
    ).toMatchSnapshot();
});

test('Custom should render with type number with attributes', () => {
    expect(render(
        <Custom max={50} min={10} onBlur={jest.fn()} onChange={jest.fn()} step={5} type="number" value={25} />)
    ).toMatchSnapshot();
});
