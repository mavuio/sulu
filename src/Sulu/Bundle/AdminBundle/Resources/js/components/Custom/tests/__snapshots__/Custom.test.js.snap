// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom should render 1`] = `
<label
  class="custom default left"
>
  <custom
    type="text"
    value="My value"
  />
</label>
`;

exports[`Custom should render append container when onClearClick callback is provided 1`] = `
<label
  class="custom default left hasAppendIcon"
>
  <custom
    type="text"
    value=""
  />
</label>
`;

exports[`Custom should render append container with icon when onClearClick callback is provided and value is set 1`] = `
<label
  class="custom default left hasAppendIcon"
>
  <custom
    type="text"
    value="test"
  />
  <div
    class="appendContainer"
  >
    <span
      aria-label="su-times"
      class="su-times clickable icon default"
      role="button"
      tabindex="0"
    />
  </div>
</label>
`;

exports[`Custom should render as headline 1`] = `
<label
  class="custom default left headline"
>
  <custom
    type="text"
    value="My value"
  />
</label>
`;

exports[`Custom should render collapsed 1`] = `
<label
  class="custom default left collapsed"
>
  <custom
    type="text"
    value=""
  />
</label>
`;

exports[`Custom should render undefined value as empty string 1`] = `
<label
  class="custom default left"
>
  <custom
    type="text"
    value=""
  />
</label>
`;

exports[`Custom should render when disabled 1`] = `
<label
  class="custom default left disabled"
>
  <custom
    disabled=""
    type="text"
    value="My value"
  />
</label>
`;

exports[`Custom should render with a character counter 1`] = `
Array [
  <label
    class="custom default left"
  >
    <custom
      type="text"
      value="asdf"
    />
  </label>,
  <label
    class="characterCounter exceeded"
  >
    -2 sulu_admin.characters_left
  </label>,
]
`;

exports[`Custom should render with a loader 1`] = `
<label
  class="custom default left"
>
  <div
    class="prependedContainer default"
  >
    <div
      class="spinner"
      style="width:20px;height:20px"
    >
      <div
        class="doubleBounce1"
      />
      <div
        class="doubleBounce2"
      />
    </div>
  </div>
  <custom
    type="text"
    value=""
  />
</label>
`;

exports[`Custom should render with a segment counter 1`] = `
Array [
  <label
    class="custom default left"
  >
    <custom
      type="text"
      value="keyword1, keyword2"
    />
  </label>,
  <label
    class="segmentCounter"
  >
    1 sulu_admin.segments_left
  </label>,
]
`;

exports[`Custom should render with autocomplete off 1`] = `
<label
  class="custom default left disabled"
>
  <custom
    autocomplete="off"
    disabled=""
    type="text"
    value="My value"
  />
</label>
`;

exports[`Custom should render with dark skin 1`] = `
<label
  class="custom dark left hasAppendIcon"
>
  <div
    class="prependedContainer dark"
  >
    <span
      aria-label="su-pen"
      class="su-pen icon dark"
    />
  </div>
  <custom
    type="text"
    value=""
  />
</label>
`;

exports[`Custom should render with icon 1`] = `
<label
  class="custom default left"
>
  <div
    class="prependedContainer default"
  >
    <span
      aria-label="su-pen"
      class="su-pen icon default"
    />
  </div>
  <custom
    type="text"
    value="My value"
  />
</label>
`;

exports[`Custom should render with custommode 1`] = `
<label
  class="custom default left"
>
  <custom
    custommode="numeric"
    type="text"
    value="My value"
  />
</label>
`;

exports[`Custom should render with invalid value 1`] = `
<label
  class="custom default left error"
>
  <custom
    type="text"
    value="My value"
  />
</label>
`;

exports[`Custom should render with placeholder 1`] = `
<label
  class="custom default left"
>
  <custom
    placeholder="My placeholder"
    type="text"
    value="My value"
  />
</label>
`;

exports[`Custom should render with type 1`] = `
<label
  class="custom default left"
>
  <custom
    type="password"
    value="My value"
  />
</label>
`;

exports[`Custom should render with type number with attributes 1`] = `
<label
  class="custom default left"
>
  <custom
    max="50"
    min="10"
    step="5"
    type="number"
    value="25"
  />
</label>
`;

exports[`Custom should render with value 1`] = `
<label
  class="custom default left"
>
  <custom
    type="text"
    value="My value"
  />
</label>
`;
